### PR TITLE
[Serializer] Fix inconsistent behaviour of nullable objects in key/value arrays

### LINKED ISSUE #45883: [Serializer] inconsistent behaviour of nullable objects in key/value arrays
### Symfony version(s) affected

6.1, 6.0, 5.4

### Description

The way nullable objects in maps (e.g. `array<string, Dummy|null>`) are denormalized is inconsistent to the behaviour when denormalizing the same value type (in this case `Dummy|null`) to a plain attribute.

### Problem 1) Object values of a map can never be `null`
* denormalizing `['foo' => ['bar' => null]]` to ` @var array<string,DummyValue|null> $foo` will always denormalize to an empty instance of `DummyValue`:  `['foo' => ['bar' => object(DummyValue)]]` (expected `['foo' => ['bar' => null]]`
* you will always end up in a `output != input` state when doing some thing like `input:json => deserialize() => serialize() => output:json` 
* though it works  (`['foo' => null]`) when it's a plain (non-map) attribute/property value (`@var DummyValue|null $foo`)
 
### Problem 2) It's impossible to use discriminated/abstract types in maps where the value is nullable
* denormalizing `['foo' => ['bar' => null]]` to `@var array<string,AbstractDummyValue|null> $foo` (where `AbstractDummyValue` has a discriminator mapping) it will throw  `NotNormalizableValueException` complaining the discriminator field is not set.
* also works as expected (`['foo' => null]`) if it's a plain (non-map) attribute/property value (`@var AbstractDummyValue|null $foo`)

### How to reproduce

```php
$loaderMock = new class() implements ClassMetadataFactoryInterface {
      public function getMetadataFor($value): ClassMetadataInterface
      {
          if (AbstractDummyValue::class === $value) {
              return new ClassMetadata(
                  AbstractDummy::class,
                  new ClassDiscriminatorMapping('type', [
                      'dummy' => DummyValue::class,
                      'another-dummy' => AnotherDummyValue::class,
                  ])
              );
          }
    
          throw new InvalidArgumentException();
      }
    
      public function hasMetadataFor($value): bool
      {
          return AbstractDummyValue::class === $value;
      }
};

$factory = new ClassMetadataFactory(new AnnotationLoader(new AnnotationReader()));
$normalizer = new ObjectNormalizer($factory, null,null, new PhpDocExtractor(), new ClassDiscriminatorFromClassMetadata($loaderMock));
$serializer = new Serializer([$normalizer, new ArrayDenormalizer()]);
$normalizer->setSerializer($serializer);
```

### Steps to reproduce 1) ("Object values of a map can never be `null`")

```php
class DummyValue 
{
}

class DummyMapOfStringToNullableObject {

    /**
     * @var array<string,DummyValue|null>
     */
    public $map;
}

// Since `$map` is nullable Id expect `$map['assertNull']` to be `null`.
$normalizedData = $serializer->denormalize([
    'map' => [
        'assertDummyMapValue' => [
            'value' => 'foo'
        ],
        'assertNull' => null,  // will be denormalized to an empty `DummyValue()`
    ],
], DummyMapOfStringToNullableObject::class);
```

... while it works actually for the following scenario:
```php
class DummyNullableObjectValue
{
    /**
     * @var DummyValue|null $object
     */
    public $object;

    /**
     * @var DummyValue|null $object
     */
    public $nullObject;
}

$normalizedData = $serializer->denormalize([
    'object' => [
        'value' => 'foo',
    ],
    'nullObject' => null, // will be denormalized as expected to `null`
], DummyNullableObjectValue::class);
```

see full tests here:
* https://github.com/phramz/symfony/blob/map-denormalization/src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php#L33

### Steps to reproduce 2) ("It's impossible to use discriminated/abstract types in maps where the value is nullable")

```php

abstract class AbstractDummyValue
{
    public $value;
}

class DummyValue extends AbstractDummyValue
{
}

class AnotherDummyValue extends AbstractDummyValue
{
}

class DummyMapOfStringToNullableAbstractObject {
    /**
     * @var array<string,AbstractDummyValue|null>
     */
    public $map;
}

// FIXME throws `NotNormalizableValueException: Type property "type" not found for the abstract object "Symfony\Component\Serializer\Tests\Normalizer\AbstractDummyValue".`
$normalizedData = $serializer->denormalize([
    'map' => [
        'assertNull' => null,
    ],
], DummyMapOfStringToNullableAbstractObject::class);

```
see full tests here:
* https://github.com/phramz/symfony/blob/map-denormalization/src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php#L60

### Possible Solution

https://github.com/symfony/symfony/pull/45884

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by phramz**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  |no
| Deprecations? |no
| Tickets       | Fix #45883
| License       | MIT
| Doc PR        |  -

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I don't know about the change, but this should target 5.4 or maybe 4.4 if the bug exists there also. Can you please confirm?

File: src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+        $this->assertArrayHasKey('assertDummyMapValue', $normalizedData->map);
+        $this->assertInstanceOf(DummyValue::class, $normalizedData->map['assertDummyMapValue']);
+
+        $this->assertArrayHasKey('assertNull', $normalizedData->map);
+
+        $this->assertNull($normalizedData->map['assertNull']);
+    }
+
+    public function testMapOfStringToAbstractNullableObject()  // TODO fix?
```
Comment (alamirault) [inline src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php]: TODO can be removed ?

File: src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        $this->assertIsArray($normalizedData->map);
+
+        $this->assertArrayHasKey('assertDummyMapValue', $normalizedData->map);
+        $this->assertInstanceOf(DummyValue::class, $normalizedData->map['assertDummyMapValue']);
+        $this->assertEquals('foo', $normalizedData->map['assertDummyMapValue']->value);
+    }
+
+    public function testMapOfStringToAbstractObjectMissingTypeAttribute()
+    {
+        $this->expectExceptionObject(new NotNormalizableValueException());
```
Comment (alamirault) [inline src/Symfony/Component/Serializer/Tests/Normalizer/MapDenormalizationTest.php]:
```suggestion
        $this->expectException(NotNormalizableValueException::class);
```

**Comment by phramz** [comment]
@nicolas-grekas yes this also applies to 5.4.x (I changed the PR to target 5.4 now) ... 4.4.x seems to be far off so running the test will result in totally different errors so i guess it's incompatible. Maybe someone with deeper knowledge/understanding of the component can check on 4.4?

**Comment by stvnwrgs** [comment]
@nicolas-grekas No one here to review? We also would love to see this fixed.

**Review by dunglas — APPROVED** [review]
LGTM. Thanks for the detailed issue and for the fix!

**Review by ogizanagi — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @phramz.

**Pull request merged**